File: cc/012.cc
// Project Euler Problem 012
//
// What is the value of the first triangle number to have over five
// hundred divisors?

#include <cstdio>
#include "primes1.h"

long int tri(long int n) {
  return n * (n+1) / 2;
}

int main(int argc, const char* argv[]) {
  Primes1<long int> primes;
  long int i = 0;
  long int t = 0;
  while (true) {
    i++;
    t = tri(i);
    if (primes.num_divisors(t) > 500) {
      break;
    }
  }
  printf("%ld\n", t);
}

File: go/problems/012.go
// Project Euler Problem 012
//
// What is the value of the first triangle number to have over five
// hundred divisors?

package problems

import (
	"fmt"

	"../primes"
)

func Problem012() string {
	var n, tri int64 = 1, 0
	for ; ; n, tri = n+1, tri+n {
		count := 1
		for _, factor := range primes.PrimeFactors(tri) {
			count *= (factor.Count + 1)
		}
		if count > 500 {
			return fmt.Sprint(tri)
		}
	}
	return ""
}

func init() {
     Functions[12] = Problem012
}
